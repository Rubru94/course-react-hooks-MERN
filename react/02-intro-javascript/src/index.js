// Copy the contents of any script from bases/ into index.js to see the execution and results.

console.log('index.js');
