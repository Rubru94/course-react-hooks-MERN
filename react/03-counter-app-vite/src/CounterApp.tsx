import { useState } from 'react'; // nomenclature --> if it starts with "use" it is a hook

export interface CounterAppProps {
  value?: number;
}

export const CounterApp = ({ value }: CounterAppProps) => {
  console.log('render'); // When status changes, the component is executed again

  const [counter, setCounter] = useState(value);

  const handleAdd = () => {
    setCounter(counter + 1);
    // setCounter((c) => c + 1);
  };

  const handleSubtract = () => setCounter(counter - 1);

  const handleReset = () => setCounter(value);

  return (
    <>
      <h1>CounterApp</h1>
      <h2> {counter} </h2>
      {/* equivalent --> it can be simplified by leaving only the reference to the function that will receive the same arguments in same order */}
      {/* <button onClick={(event) => handleAdd(event)}>+1</button> */}
      <button onClick={handleAdd}> +1 </button>
      <button onClick={handleSubtract}> -1 </button>
      <button onClick={handleReset}> Reset </button>

      {/* <button onClick={(event) => handleAdd(event, 1)}>+1</button> */}
    </>
  );
};
