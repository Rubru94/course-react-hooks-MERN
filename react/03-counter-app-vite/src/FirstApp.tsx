const objMessage = {
  message: 'Hello world !!',
  title: 'Greeting',
};

const getMessage = (msg?: string): string => {
  return msg ?? objMessage.message;
};

const getResult = (a: number, b: number): number => {
  return a + b;
};

export const FirstApp = () => {
  return (
    <>
      <h1>First app</h1>
      <h2>Message: {getMessage('Hi')}</h2>
      <h3>Result: {getResult(2, 5)}</h3>
      {/* You cannot print objects directly in react (you can use JSON.stringify), nor async function results. */}
      {/* <code>{JSON.stringify(objMessage)}</code> */}
      <p>Subtitle</p>
    </>
  );
};
