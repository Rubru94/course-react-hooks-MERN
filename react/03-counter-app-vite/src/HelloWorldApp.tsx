export function App() {
  // document.createElement...
  return <h1>Hello World !!!</h1>;
}

// rfc + tab
export const HelloWorldApp = () => {
  return <h1>HelloWorldApp</h1>;
};

export const App2 = () => <h1>Hello World!!! from arrow function</h1>;

/** default export
 * 
    function App() {
      // document.createElement...
      return <h1>Hello World !!!</h1>;
    }

    export default App;
 * 
 */
