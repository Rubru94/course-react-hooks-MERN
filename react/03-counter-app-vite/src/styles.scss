/* global styles */

html,
body {
  background-color: darkcyan;
  color: white;
  font-family: Helvetica, Arial, sans-serif;
  font-size: 1.3rem;
  padding: 70px;
}
