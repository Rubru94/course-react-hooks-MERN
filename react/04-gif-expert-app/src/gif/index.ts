export * from './GifGrid';
export * from './GifItem';
export * from './gif.service';
export * from './hooks/useFetchGifs.hook';
