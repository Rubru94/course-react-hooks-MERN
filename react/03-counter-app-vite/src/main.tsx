import React from 'react';
import ReactDOM from 'react-dom/client';

import { CounterApp } from './CounterApp';

import './styles.scss';

ReactDOM.createRoot(document.getElementById('root') as HTMLElement).render(
  <React.StrictMode>
    <CounterApp value={30}></CounterApp>
  </React.StrictMode>,
);
