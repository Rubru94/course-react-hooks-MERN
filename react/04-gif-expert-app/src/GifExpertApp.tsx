import { useState } from 'react';
import { AddCategory } from './category/AddCategory';
import { GifGrid } from './gif/GifGrid';

export const GifExpertApp = () => {
  // useContextSnippet + Tab
  const [categories, setCategories] = useState(['Metal Gear Solid']);

  // Hooks must not be created according to conditions
  /* if (true) {
    const [categoriesNew, setCategoriesNew] = useState(['Metal Gear Solid']);
  } */

  const isValidCategory = (category: string): boolean => {
    return categories
      .map((c) => c.toLowerCase())
      .includes(category.toLowerCase());
  };

  const onAddCategory = (newCategory: string) => {
    if (isValidCategory(newCategory)) return;

    setCategories([newCategory, ...categories]);
    // equivalent
    /* const onAddCategory = () => setCategories((cat) => [...cat, 'foo']); */ // with callback
  };

  return (
    <>
      <h1>GifExpertApp</h1>

      <AddCategory
        onNewCategory={(value: string) => onAddCategory(value)}
      ></AddCategory>

      {/* <button onClick={onAddCategory}>Add</button> */}
      {/* <button onClick={(event) => onAddCategory(event)}>Add</button> */}

      {categories.map((category) => (
        <GifGrid key={category} category={category} />
      ))}
    </>
  );
};
