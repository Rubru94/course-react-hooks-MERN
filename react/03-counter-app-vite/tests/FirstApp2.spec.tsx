import { render, screen } from '@testing-library/react';
import { FirstApp } from '../src/FirstApp';
import React from 'react';

describe('FirstApp Component - Improved test', () => {
  const title = 'Helloooo';
  const subtitle = 'My subtitle';

  it('should match with snapshot', () => {
    const { container } = render(<FirstApp title={title} num={0} />);

    expect(container).toMatchSnapshot();
  });

  it('should show message "Helloooo"', () => {
    render(<FirstApp title={title} num={0} />);

    // screen.debug(); --> print object
    expect(screen.getByText(title)).toBeTruthy();
  });

  it('should show title into h2 tag', () => {
    render(<FirstApp title={title} num={0} />);

    // getByRole --> get by html tag --> this case h2
    expect(screen.getByRole('heading', { level: 2 }).innerHTML).toContain(title);
  });

  it('should show subtitle from props', () => {
    render(<FirstApp title={title} subtitle={subtitle} num={0} />);

    expect(screen.getAllByText(subtitle).length).toBeGreaterThan(0);
  });
});
