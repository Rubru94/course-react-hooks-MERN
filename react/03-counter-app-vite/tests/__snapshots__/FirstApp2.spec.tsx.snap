// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FirstApp Component should match with snapshot 1`] = `
<div>
  <h1>
    First app
  </h1>
  <h2
    data-testid="test-title"
  >
    Helloooo
  </h2>
  <p>
    default
  </p>
  <p>
    0
  </p>
</div>
`;
